pkg/resources/nlb_service/datasource_list.go

```diff
@@ -0,0 +1,289 @@
+	if resp.Diagnostics.HasError() {
+		return
+	}
+	ctx, cancel := context.WithTimeout(ctx, t)
+	defer cancel()
+
+	ctx = exoapi.WithEndpoint(ctx, exoapi.NewReqEndpoint(d.env, data.Zone.ValueString()))
+
+	var x string
+	switch {
+	case !data.NLBID.IsNull():
+		x = data.NLBID.ValueString()
+	case !data.NLBName.IsNull():
+		x = data.NLBName.ValueString()
+	default:
+		resp.Diagnostics.AddError(
+			"Either nlb_name or nlb_id must be specified",
+			"",
+		)
+		return
+	}
+
+	nlb, err := d.client.FindNetworkLoadBalancer(ctx, data.Zone.ValueString(), x)
+	if err != nil {
+		resp.Diagnostics.AddError(
+			"Unable to find Network Load Balancer",
+			err.Error(),
+		)
+		return
+	}
+
+	data.NLBName = types.StringValue(*nlb.Name)
+
+	for _, service := range nlb.Services {
+		ctx = tflog.SetField(ctx, "exoscale_service", service)
+		tflog.Info(ctx, "Loop NLB service")
+
+		var serviceState Service
+
+		if service.Description != nil {
+			serviceState.Description = types.StringValue(*service.Description)
+		}
+		if service.ID != nil {
+			serviceState.ID = types.StringValue(*service.ID)
+		}
+		if service.InstancePoolID != nil {
+			serviceState.InstancePoolID = types.StringValue(*service.InstancePoolID)
+		}
+		if service.Name != nil {
+			serviceState.Name = types.StringValue(*service.Name)
+		}
+		if service.Port != nil {
+			serviceState.Port = types.Int64Value(int64(*service.Port))
+		}
+		if service.Protocol != nil {
+			serviceState.Protocol = types.StringValue(*service.Protocol)
+		}
+		if service.State != nil {
+			serviceState.State = types.StringValue(*service.State)
+		}
+		if service.Strategy != nil {
+			serviceState.Strategy = types.StringValue(*service.Strategy)
+		}
+		if service.TargetPort != nil {
+			serviceState.TargetPort = types.Int64Value(int64(*service.TargetPort))
+		}
+
+		var healtcheckState Healthcheck
+
+		if service.Healthcheck.Interval != nil {
+			healtcheckState.Interval = types.Int64Value(int64(*service.Healthcheck.Interval / time.Second))
+		}
+		if service.Healthcheck.Mode != nil {
+			healtcheckState.Mode = types.StringValue(*service.Healthcheck.Mode)
+		}
+		if service.Healthcheck.Port != nil {
+			healtcheckState.Port = types.Int64Value(int64(*service.Healthcheck.Port))
+		}
+		if service.Healthcheck.Retries != nil {
+			healtcheckState.Retries = types.Int64Value(int64(*service.Healthcheck.Retries))
+		}
+		if service.Healthcheck.TLSSNI != nil {
+			healtcheckState.TLSSNI = types.StringValue(*service.Healthcheck.TLSSNI)
+		}
+		if service.Healthcheck.Timeout != nil {
+			healtcheckState.Timeout = types.Int64Value(int64(*service.Healthcheck.Timeout / time.Second))
+		}
+		if service.Healthcheck.URI != nil {
+			healtcheckState.URI = types.StringValue(*service.Healthcheck.URI)
+		}
+
+		serviceState.Healthcheck = healtcheckState
+
+		data.NLBServiceList = append(data.NLBServiceList, serviceState)
+	}
+
+	resp.Diagnostics.Append(resp.State.Set(ctx, &data)...)
+	if resp.Diagnostics.HasError() {
```

Comment:
That's what Terraform tutorial shows: https://developer.hashicorp.com/terraform/tutorials/providers-plugin-framework/providers-plugin-framework-data-source-read#implement-read-functionality

Is it not possible the line above might generate an error which wouldn't be raised without it ?

Reply:
I don't see how it would make any difference, on error it just calls `return` to interrupt the flow but there is no code after it that would execute. Tutorial probably expects you would extend it and add more code after it.

Here is an example from hashicorp themselves that doesn't check for error on the last line:
https://github.com/hashicorp/terraform-provider-scaffolding-framework/blob/main/internal/provider/example_data_source.go#L104

Reply:
Thanks for the confirmation, I'll remove it.